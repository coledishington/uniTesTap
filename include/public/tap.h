#ifndef __TAP_H__
#define __TAP_H__

/**
 * @var TAP
 *
 * A tap handle for use with all tap functions.
 */
typedef struct TAP TAP;

/**
 * @var test_t
 *
 * Type of TAP test function.
 */
typedef int (*test_t)(void);

/**
 * @fn tap_init
 *
 * Allocates and returns a TAP handle that should be passed to tap functions.
 *
 * @param d_tap double pointer to TAP handle. If tap function calls do not
 *              need to be re-entrant, NULL can be passed.
 *
 * @return 0 on success, errno-like value otherwise.
 */
int tap_init(TAP **d_tap);

/**
 * @fn tap_cleanup
 *
 * @param tap a TAP handle allocated by tap_init().
 *
 * Cleanup any resources allocated by libtap associated with the passed
 * handle.
 */
void tap_cleanup(TAP *tap);

/**
 * @fn tap_register
 *
 * Register a test to run in tap_runall().
 *
 * @param tap a tap handle allocated by tap_init().
 * @param test the test function to register.
 * @param description an optional description of the registered test.
 *
 * @return 0 on success, errno-like value otherwise.
 */
int tap_register(TAP *tap, test_t test, const char *description);

/**
 * @fn tap_runall
 *
 * Run all tests registered in the global TAP tests list.
 *
 * @param tap a tap handle allocated by tap_init().
 *
 * @return 0 on success, errno-like value otherwise.
 */
int tap_runall(TAP *tap);

/**
 * @fn tap_runall_and_cleanup
 *
 * Run all tests registered in the global TAP tests list and cleanup.
 *
 * @param tap a tap handle allocated by tap_init().
 *
 * @return 0 on success, errno-like value otherwise.
 */
static inline int tap_runall_and_cleanup(TAP *tap) {
    int ret;

    ret = tap_runall(tap);
    tap_cleanup(tap);
    return ret;
}

#endif /* __TAP_H__ */