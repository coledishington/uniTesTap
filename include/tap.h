#ifndef __TAP_H__
#define __TAP_H__

/**
 * @var test_t
 *
 * Type of TAP test function.
 */
typedef int (*test_t)(void);

/**
 * @fn tap_register
 *
 * Register a test to run in tap_runall().
 *
 * @return 0 on success, errno-like value otherwise.
 */
int tap_register(test_t test, const char *description);

/**
 * @fn tap_runall
 *
 * Run all tests registered in the global TAP tests list.
 *
 * @return 0 on success, errno-like value otherwise.
 */
int tap_runall(void);

#endif /* __TAP_H__ */