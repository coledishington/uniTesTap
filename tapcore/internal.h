#ifndef __INTERNAL_H__
#define __INTERNAL_H__
#include <poll.h>
#include <stdbool.h>
#include <tapstruct.h>
#include <taptest.h>

struct test_run {
    struct test test;
    tap_cmd_t *cmd;
    pid_t pid;
    int outfd;
    int exitstatus;
};

int tap_start_testrun(struct test *test, struct test_run *testrun);

int tap_wait_for_testrun(struct test_run *testrun);

void tap_cleanup_testrun(struct test_run *testrun);

#endif /* __INTERNAL_H__ */
