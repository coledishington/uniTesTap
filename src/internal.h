#ifndef __TAP_INTERNAL_H__
#define __TAP_INTERNAL_H__
#include <stdbool.h>
#include <stdint.h>
#include <tap.h>

#define ARRAY_LEN(A) (sizeof(A) / sizeof(*A))

#define _STRING_JOIN(expr1, expr2) (expr1)##(expr2)
#define STRING_JOIN(expr1, expr2) _STRING_JOIN(expr1, expr2)

#define STATIC_ASSERT_ID STRING_JOIN(static_assert_, __LINE__)
#define STATIC_ASSERT_TEST(expr) ((expr) ? 1 : -1)

/* Static assert that uses both the bitmask and array tricks */
#define STATIC_ASSERT(expr)                              \
    typedef struct {                                     \
        int STATIC_ASSERT_ID : STATIC_ASSERT_TEST(expr); \
    } STATIC_ASSERT_ID[STATIC_ASSERT_TEST(expr)]

#define TAP_DIRECTIVE_SKIP "SKIP"
#define TAP_DIRECTIVE_TODO "TODO"
#define TAP_PIPE_RX 0
#define TAP_PIPE_TX 1

struct test {
    char *description;
    test_t funct;
    size_t id;
};

struct tap_string;
typedef struct tap_string tap_string_t;

int tap_pipe_setup(int fds[2]);

tap_string_t *tap_string_ctor(const char *str);

int tap_string_concat(tap_string_t *tstr, const char *str);

int tap_string_concat_printf(tap_string_t *tstr, const char *fmt, ...);

char *tap_string_dtor(tap_string_t *tstr, bool free_str);

void tap_print_testpoint(bool success, struct test *test,
                         const char *directive);

void tap_print_internal_error(int err, struct test *test, const char *reason);

#endif /* __TAP_INTERNAL_H__ */
