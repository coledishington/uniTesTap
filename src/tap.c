#include <assert.h>
#include <errno.h>
#include <signal.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/types.h>
#include <sys/wait.h>
#include <tap.h>
#include <unistd.h>

#define MAX_TESTS 10

struct cfg {
    size_t n_tests;
    test_t tests[MAX_TESTS];
};

struct cfg cfg = {
    .n_tests = 0,
};

static void tap_report_test(size_t test_id, int wres) {
    int res;

    if (WIFEXITED(wres)) {
        res = WEXITSTATUS(wres);
        switch (res) {
            case 0:
                printf("ok %zu\n", test_id);
                break;
            default:
                printf("not ok %zu\n", test_id);
                break;
        }
    } else if (WIFSIGNALED(wres)) {
        int sig = WTERMSIG(wres);
        const char *sig_name = strsignal(sig);

        if (!sig_name) {
            sig_name = "UNKNOWN";
        }
        printf("not ok %zu # test terminated via %s(%d)\n", test_id, sig_name,
               sig);
    } else {
        printf("no ok %zu # test exited for unknown reason\n", test_id);
    }
}

static void tap_run_test_and_exit(test_t test) {
    int res;

    res = test();
    _exit(res);
}

static int tap_evaluate(size_t test_id, test_t test) {
    int wres;
    pid_t cpid;

    cpid = fork();
    if (cpid == 0) {
        /* Child process will run test and exit */
        tap_run_test_and_exit(test);
        _exit(-1);
    }
    if (cpid == -1) {
        /* Child process spawning failed */
        return errno;
    }

    /* Wait for child to exit */
    cpid = waitpid(cpid, &wres, 0);
    if (cpid < 0) {
        /* Child process is lost */
        return errno;
    }

    /* Report test success */
    tap_report_test(test_id, wres);
    return 0;
}

void tap_register(test_t test) {
    assert(cfg.n_tests + 1 < MAX_TESTS);

    cfg.tests[cfg.n_tests] = test;
    cfg.n_tests++;
}

int tap_runall(void) {
    int err = 0;

    printf("1..%zu\n", cfg.n_tests);
    for (size_t test_idx = 0; test_idx < cfg.n_tests; test_idx++) {
        test_t test = cfg.tests[test_idx];
        size_t test_id = test_idx + 1;
        err = tap_evaluate(test_id, test);
        if (err != 0) {
            printf("Bail out! internal test runner error %s(%d): ",
                   strerror(err), err);
            break;
        }
    }
    return err;
}
