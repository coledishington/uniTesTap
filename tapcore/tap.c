#include <assert.h>
#include <errno.h>
#include <signal.h>
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/types.h>
#include <sys/wait.h>
#include <tap.h>
#include <tapio.h>
#include <tapstruct.h>
#include <taptest.h>
#include <unistd.h>

#define MAX_TESTS 200

struct cfg {
    size_t n_tests;
    struct test tests[MAX_TESTS];
};

struct cfg cfg = {
    .n_tests = 0,
};

static void tap_report_test(struct test *test, int wres,
                            const char *directive) {
    bool passed = false;

    if (WIFEXITED(wres)) {
        passed = WEXITSTATUS(wres) == 0;
    } else if (WIFSIGNALED(wres)) {
        const char *sig_name;
        int sig;

        sig = WTERMSIG(wres);
        sig_name = strsignal(sig);
        if (!sig_name) {
            sig_name = "UNKNOWN";
        }
        printf("# test terminated via %s(%d)\n", sig_name, sig);
    } else {
        printf("# test %zu exited for unknown reason\n", test->id);
    }
    tap_print_testpoint(passed, test, directive);
}

static int tap_parse_directive(const char *line, char **d_directive) {
    const char *newline;
    char *directive;
    size_t cmd_len;

    if (strncasecmp(":" TAP_DIRECTIVE_SKIP, line, sizeof(TAP_DIRECTIVE_SKIP)) !=
            0 &&
        strncasecmp(":" TAP_DIRECTIVE_TODO, line, sizeof(TAP_DIRECTIVE_TODO)) !=
            0) {
        return 0;
    }

    /* Copy everything before the first newline */
    newline = strchr(line, '\n');
    if (newline) {
        cmd_len = newline - line;
    } else {
        cmd_len = strlen(line);
    }

    /* Skip past the ':' */
    directive = strndup(line + 1, cmd_len - 1);
    if (!directive) {
        return errno;
    }

    *d_directive = directive;
    return 0;
}

static int tap_process_test_output(int test_fd, char **d_directive) {
    char *directive = NULL;
    size_t line_len = 0;
    char *line = NULL;
    ssize_t bytes;
    FILE *test_fp;

    test_fp = fdopen(test_fd, "r");
    for (; (bytes = getline(&line, &line_len, test_fp)) != -1;) {
        char *ln_directive = NULL;
        int err;

        if (bytes == 0 || *line == '\n') {
            continue;
        }
        err = tap_parse_directive(line, &ln_directive);
        if (err != 0) {
            break;
        }
        if (!ln_directive) {
            /* Debug from the test, output as TAP comment */
            printf("# %s\n", line);
            continue;
        }
        if (directive) {
            /* Only allow one directive command per test, warn the extra is
             * ignored */
            printf("# One directive command per test: ignoring '%s'\n",
                   ln_directive);
            free(ln_directive);
        }
        directive = ln_directive;
    }
    if (directive) {
        *d_directive = directive;
    }
    free(line);
    return 0;
}

static void tap_run_test_and_exit(struct test *test) {
    int res;

    res = test->funct();
    fflush(NULL);
    _exit(res);
}

static int tap_evaluate(struct test *test) {
    int pipefd[2] = {-1, -1};
    char *directive = NULL;
    int wres, err;
    pid_t cpid;

    /* Communicate fail condition on pipe */
    err = tap_pipe_setup(pipefd);
    if (err != 0) {
        tap_print_internal_error(err, test, "failed to create pipe");
        return err;
    }

    /* Flush stdout and stderr to avoid child duplicating buffered output */
    fflush(NULL);
    cpid = fork();
    if (cpid == 0) {
        close(pipefd[TAP_PIPE_RX]);
        dup2(pipefd[TAP_PIPE_TX], STDOUT_FILENO);
        dup2(pipefd[TAP_PIPE_TX], STDERR_FILENO);
        close(pipefd[TAP_PIPE_TX]);
        /* Child process will run test and exit */
        tap_run_test_and_exit(test);
        _exit(-1);
    }
    if (cpid == -1) {
        err = errno;
        close(pipefd[TAP_PIPE_RX]);
        close(pipefd[TAP_PIPE_TX]);
        tap_print_internal_error(err, test, "failed to fork process");
        return err;
    }
    close(pipefd[TAP_PIPE_TX]);

    err = tap_process_test_output(pipefd[TAP_PIPE_RX], &directive);
    if (err != 0) {
        free(directive);
        return err;
    }
    close(pipefd[TAP_PIPE_RX]);

    /* Wait for child to exit */
    cpid = waitpid(cpid, &wres, 0);
    if (cpid < 0) {
        /* Child process is lost */
        return errno;
    }

    /* Report test success */
    tap_report_test(test, wres, directive);
    free(directive);
    return 0;
}

int tap_register(test_t funct, const char *in_description) {
    char *description = NULL;

    assert(cfg.n_tests + 1 < MAX_TESTS);

    if (in_description) {
        description = strdup(in_description);
        if (!description) {
            return errno;
        }
    }

    cfg.tests[cfg.n_tests] = (struct test){
        .id = cfg.n_tests + 1,
        .funct = funct,
        .description = description,
    };
    cfg.n_tests++;
    return 0;
}

int tap_runall(void) {
    int err = 0;

    printf("1..%zu\n", cfg.n_tests);
    for (size_t idx = 0; idx < cfg.n_tests; idx++) {
        struct test *test;
        int err;

        test = &cfg.tests[idx];
        err = tap_evaluate(test);
        if (err != 0) {
            printf("Bail out! internal test runner error %s(%d): ",
                   strerror(err), err);
            break;
        }
    }
    return err;
}
