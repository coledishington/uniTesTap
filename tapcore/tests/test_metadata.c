#include <tap.h>

#include "internal.h"

int main(void) {
    tap_register(pass, "Test description");
    tap_register(fail, "Test description");
    tap_runall();
}
