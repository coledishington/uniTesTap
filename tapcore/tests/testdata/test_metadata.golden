1..2
ok 1 - Test description
not ok 2 - Test description
